superset/assets/src/visualizations/core/components/SuperChart.jsx

```diff
@@ -0,0 +1,169 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import getChartComponentRegistry from '../registries/ChartComponentRegistrySingleton';
+import getChartTransformPropsRegistry from '../registries/ChartTransformPropsRegistrySingleton';
+
+const STATUS = {
+  IDLE: 1,
+  LOADING: 2,
+  SUCCESS: 3,
+  FAILURE: 4,
+};
+const IDENTITY = x => x;
+
+const propTypes = {
+  id: PropTypes.string,
+  className: PropTypes.string,
+  chartProps: PropTypes.object.isRequired,
+  chartType: PropTypes.string.isRequired,
+  preTransformProps: PropTypes.func,
+  overrideTransformProps: PropTypes.func,
+  postTransformProps: PropTypes.func,
+  onRenderSuccess: PropTypes.func,
+  onRenderFailure: PropTypes.func,
+};
+const defaultProps = {
+  id: '',
+  className: '',
+  preTransformProps: IDENTITY,
+  overrideTransformProps: undefined,
+  postTransformProps: IDENTITY,
+  onRenderSuccess() {},
+  onRenderFailure() {},
+};
+
+class SuperChart extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.state = {
+      status: STATUS.IDLE,
+      error: null,
+      Renderer: null,
+      transformProps: null,
+    };
+  }
+
+  componentDidMount() {
+    this.mounted = true;
+    this.loadChartType(this.props);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const { chartType, overrideTransformProps } = this.props;
+    if (nextProps.chartType !== chartType
+      || nextProps.overrideTransformProps !== overrideTransformProps
+    ) {
+      this.loadChartType(nextProps);
+    }
+  }
+
+  componentDidUpdate() {
+    const { onRenderSuccess, onRenderFailure } = this.props;
+    const { status, error } = this.state;
+    if (status === STATUS.SUCCESS) {
+      onRenderSuccess();
+    } else if (status === STATUS.FAILURE) {
+      onRenderFailure(error);
+    }
+  }
+
+  componentWillUnmount() {
+    this.mounted = false;
+  }
+
+  loadChartType({ chartType, overrideTransformProps }) {
+    if (this.mounted) {
+      // Clear state
+      this.setState({
+        status: chartType ? STATUS.IDLE : STATUS.LOADING,
+        error: null,
+        Renderer: null,
+        transformProps: null,
+      });
+
+      if (chartType) {
+        const componentPromise = getChartComponentRegistry().getAsPromise(chartType);
+        const transformPropsPromise = overrideTransformProps
+          ? Promise.resolve(overrideTransformProps)
+          : getChartTransformPropsRegistry().getAsPromise(chartType);
+
+        Promise.all([componentPromise, transformPropsPromise])
+          .then(
+            // on success
+            ([Renderer, transformProps]) => {
+              if (this.mounted) {
+                this.setState({
+                  status: STATUS.SUCCESS,
+                  Renderer: Renderer.default || Renderer,
```

Comment:
could we keep the comment about this being due to the chart module export type? I think it's sort of confusing/wouldn't be as obvious with the chart loader wrapper ... I think at some point it'd be ideal if all modules were the same type.

Reply:
Add comments